ccl/streamingccl/streamingest: TestTenantStreamingCreationErrors failed

ccl/streamingccl/streamingest.TestTenantStreamingCreationErrors [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/13018282?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/13018282?buildTab=artifacts#/) on master @ [da982d571a4822164beeaa3c33e61f677c80a9fa](https://github.com/cockroachdb/cockroach/commits/da982d571a4822164beeaa3c33e61f677c80a9fa):

Fatal error:

```
panic: test timed out after 14m57s
running tests:
	TestTenantStreamingCreationErrors (14m24s)

```
Stack: 

```
goroutine 62744 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x3b9
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x2d
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x79f29f8, 0xc0178a0360})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:392 +0x94
* runtime/pprof.Do({0x79f26e8?, 0xb9288a0?}, {{0xc001180240?, 0x11582c0?, 0xc00ea36870?}}, 0xc0047f97a8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x9d
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0xc0047f9770?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:390 +0x52
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 62271
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:382 +0xe5
* 
* goroutine 62341 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x79f29f8, 0xc0178a0480})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:392 +0x94
* runtime/pprof.Do({0x79f26e8?, 0xb9288a0?}, {{0xc001180240?, 0x11582c0?, 0xc00ea36870?}}, 0xc010b95fa8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x9d
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0xc010b95f70?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:390 +0x52
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 62271
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:382 +0xe5
* 
* goroutine 62337 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x79f29f8, 0xc001a1e5d0})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:392 +0x94
* runtime/pprof.Do({0x79f26e8?, 0xb9288a0?}, {{0xc001180240?, 0x11582c0?, 0xc00ea36870?}}, 0xc0047f57a8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x9d
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0xc0047f5770?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:390 +0x52
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 62271
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:382 +0xe5
* 
* goroutine 62338 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x79f29f8, 0xc001a1e030})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:392 +0x94
* runtime/pprof.Do({0x79f26e8?, 0xb9288a0?}, {{0xc001180240?, 0x11582c0?, 0xc00ea36870?}}, 0xc0047f9fa8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x9d
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0xc0047f9f70?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:390 +0x52
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 62271
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:382 +0xe5
* 
* goroutine 62330 [chan receive, 2 minutes]:
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop.func1({0x79f29f8, 0xc00aa11860})
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:392 +0x94
* runtime/pprof.Do({0x79f26e8?, 0xb9288a0?}, {{0xc001180240?, 0x11582c0?, 0xc00ea36870?}}, 0xc004f93fa8)
* 	GOROOT/src/runtime/pprof/runtime.go:51 +0x9d
* github.com/cockroachdb/pebble.(*tableCacheShard).releaseLoop(0xc004f93f70?)
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:390 +0x52
* created by github.com/cockroachdb/pebble.(*tableCacheShard).init in goroutine 62271
* 	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/table_cache.go:382 +0xe5
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreamingCreationErrors.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvtenant/setting_overrides.go
+++ b/pkg/kv/kvclient/kvtenant/setting_overrides.go
@@ -72,7 +72,14 @@ func (c *connector) runTenantSettingsSubscription(ctx context.Context, startupCh
 				err := errors.DecodeError(ctx, e.Error)
 				log.Errorf(ctx, "error consuming TenantSettings RPC: %v", err)
 				if startupCh != nil && errors.Is(err, &kvpb.MissingRecordError{}) && c.earlyShutdownIfMissingTenantRecord {
-					startupCh <- err
+					select {
+					case startupCh <- err:
+					case <-ctx.Done():
+						// Shutdown or cancellation may prevent the receiver from getting
+						// this error, so short circuit.
+						return
+					}
+
 					close(startupCh)
 					c.tryForgetClient(ctx, client)
 					return
@@ -131,7 +138,13 @@ func (c *connector) runTenantSettingsSubscription(ctx context.Context, startupCh
 					log.Infof(ctx, "received initial tenant settings")
 
 					if startupCh != nil {
-						startupCh <- nil
+						select {
+						case startupCh <- nil:
+						case <-ctx.Done():
+							// Shutdown or cancellation may prevent the receiver from getting
+							// this completion message, so short circuit.
+							return
+						}
 						close(startupCh)
 						startupCh = nil
 					}

